### PR TITLE
GG-36605 Add date to JavaLoggerFormatter message layout

### PR DESCRIPTION
**Pull request by solveme**

### CODE REVIEW

File: modules/core/src/main/java/org/apache/ignite/internal/util/IgniteUtils.java [hunk 1]
```diff
@@ -557,6 +557,10 @@ public abstract class IgniteUtils {
     private static final DateTimeFormatter THREAD_DUMP_FMT =
         DateTimeFormatter.ofPattern("yyyy/MM/dd HH:mm:ss z").withZone(ZoneId.systemDefault());
 
+    /** ISO-8601 compliant date-time format. */
+    public static final DateTimeFormatter ISO_DATE_FMT =
+        DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm:ss,SSS").withZone(ZoneId.systemDefault());
```
Comment (sashapolo) [inline modules/core/src/main/java/org/apache/ignite/internal/util/IgniteUtils.java:562]: can we use `DateTimeFormatter.ISO_LOCAL_DATE_TIME` instead?

File: modules/core/src/test/java/org/apache/ignite/logger/java/JavaLoggerFormatterTest.java [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+            Thread.currentThread().getName() + "][" +
+            category + "] " +
+            message;
+    }
+
+    static void assertSimilar(String expected, String actual) {
+        assertTrue("\nExpected: " + expected + "\nActual:   " + actual, actual.startsWith(expected));
+    }
+
+}
```
Comment (sashapolo) [inline modules/core/src/test/java/org/apache/ignite/logger/java/JavaLoggerFormatterTest.java:178]: Missing line break, please check your IDE settings

File: modules/core/src/test/java/org/apache/ignite/logger/java/JavaLoggerFormatterTest.java [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+
+        // then
+        assertSimilar(buildExpectedString(msg, expectedCategory), result);
+    }
+
+    void assertFormattedOutput(String category, String expectedCategory) {
+        assertFormattedOutput("Power overwhelming", category, expectedCategory);
+    }
+
+    static final DateTimeFormatter ISO_DATE_FORMAT =
```
Comment (sashapolo) [inline modules/core/src/test/java/org/apache/ignite/logger/java/JavaLoggerFormatterTest.java:149]: What is this variable for?

File: modules/core/src/test/java/org/apache/ignite/logger/java/JavaLoggerFormatterTest.java [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+ * Java log formatter test.
+ */
+@GridCommonTest(group = "Logger")
+public class JavaLoggerFormatterTest {
+
+    private static JavaLoggerFormatter formatter;
+
+    @BeforeClass
+    public static void beforeClass() throws Exception {
+        formatter = new JavaLoggerFormatter();
```
Comment (sashapolo) [inline modules/core/src/test/java/org/apache/ignite/logger/java/JavaLoggerFormatterTest.java:43]: what's the point of having this method? `formatter` field can be initialized directly in the declaration

File: modules/core/src/test/java/org/apache/ignite/logger/java/JavaLoggerFormatterTest.java [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+        System.setProperty("IGNITE_JAVA_LOGGER_DATE_FORMAT", "foobar");
+
+        // when
+        DateTimeFormatter fmt = JavaLoggerFormatter.resolveDateFormat();
+
+        // then
+        assertEquals(BASE_DATE, fmt.format(Instant.ofEpochMilli(BASE_MILLIS)));
+    }
+
+    void assertFormattedOutput(String msg, String category, String expectedCategory) {
```
Comment (sashapolo) [inline modules/core/src/test/java/org/apache/ignite/logger/java/JavaLoggerFormatterTest.java:134]: Please make all helper methods in this class `private`

File: modules/core/src/main/java/org/apache/ignite/logger/java/JavaLoggerFormatter.java [hunk 6]
```diff
@@ -19,47 +19,72 @@
...
-    public static final String ANONYMOUS_LOGGER_NAME = "UNKNOWN";
+    static final String ANONYMOUS_LOGGER_NAME = "UNKNOWN";
+
+    /** Default date-time format. */
+    static final DateTimeFormatter DEFAULT_DATE_FMT = IgniteUtils.ISO_DATE_FMT;
+
+    /** Log message date-time formatter. */
+    static final DateTimeFormatter DATE_FMT = resolveDateFormat();
+
+    /** Property for overriding default dat-time format. */
```
Comment (sashapolo) [inline modules/core/src/main/java/org/apache/ignite/logger/java/JavaLoggerFormatter.java:44]:
```suggestion
    /** Property for overriding default date-time format. */
```

File: modules/core/src/main/java/org/apache/ignite/logger/java/JavaLoggerFormatter.java [hunk 7]
```diff
@@ -19,47 +19,72 @@
...
+    static final String IGNITE_JUL_DATE_FORMAT_PROPERTY = "IGNITE_JAVA_LOGGER_DATE_FORMAT";
 
     /** {@inheritDoc} */
     @Override public String format(LogRecord record) {
         String threadName = Thread.currentThread().getName();
 
         String logName = record.getLoggerName();
 
-        if (logName == null)
+        if (logName == null || logName.equals(""))
```
Comment (sashapolo) [inline modules/core/src/main/java/org/apache/ignite/logger/java/JavaLoggerFormatter.java:53]: `logName.isEmpty()`?

File: modules/core/src/main/java/org/apache/ignite/logger/java/JavaLoggerFormatter.java [hunk 8]
```diff
@@ -19,47 +19,72 @@
...
+    static final String ANONYMOUS_LOGGER_NAME = "UNKNOWN";
+
+    /** Default date-time format. */
+    static final DateTimeFormatter DEFAULT_DATE_FMT = IgniteUtils.ISO_DATE_FMT;
+
+    /** Log message date-time formatter. */
+    static final DateTimeFormatter DATE_FMT = resolveDateFormat();
+
+    /** Property for overriding default dat-time format. */
+    static final String IGNITE_JUL_DATE_FORMAT_PROPERTY = "IGNITE_JAVA_LOGGER_DATE_FORMAT";
```
Comment (sashapolo) [inline modules/core/src/main/java/org/apache/ignite/logger/java/JavaLoggerFormatter.java:45]: How is this property related to the issue described in the ticket? What problem are we solving here?

File: modules/core/src/test/java/org/apache/ignite/logger/java/JavaLoggerFormatterTest.java [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+        return "[" +
+            BASE_DATE + "][" +
+            Level.INFO + "][" +
+            Thread.currentThread().getName() + "][" +
+            category + "] " +
+            message;
+    }
+
+    static void assertSimilar(String expected, String actual) {
+        assertTrue("\nExpected: " + expected + "\nActual:   " + actual, actual.startsWith(expected));
```
Comment (sashapolo) [inline modules/core/src/test/java/org/apache/ignite/logger/java/JavaLoggerFormatterTest.java:175]: Can be written shorter: `assertThat(actual, startsWith(expected));`

File: modules/core/src/test/java/org/apache/ignite/logger/java/JavaLoggerFormatterTest.java [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+    static String buildExpectedString(@NotNull String message, @Nullable String category) {
+        return "[" +
+            BASE_DATE + "][" +
+            Level.INFO + "][" +
+            Thread.currentThread().getName() + "][" +
+            category + "] " +
+            message;
+    }
+
+    static void assertSimilar(String expected, String actual) {
```
Comment (sashapolo) [inline modules/core/src/test/java/org/apache/ignite/logger/java/JavaLoggerFormatterTest.java:174]: I'd propose to rename this method to `assertStartsWith`

File: modules/core/src/test/java/org/apache/ignite/logger/java/JavaLoggerFormatterTest.java [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+    @Test
+    public void testThrowable() {
+        // given
+        final String msg = "Greed is good";
+        final String category = "probe";
+        final String exceptionMsg = "Not enough minerals";
+
+        LogRecord record = buildLogRecord(msg, category);
+        try {
+            throw new RuntimeException(exceptionMsg);
```
Comment (sashapolo) [inline modules/core/src/test/java/org/apache/ignite/logger/java/JavaLoggerFormatterTest.java:90]: Why do you need to "throw" this exception? What's wrong with `record.setThrown(new RuntimeException(exceptionMsg)`?

File: modules/core/src/test/java/org/apache/ignite/logger/java/JavaLoggerFormatterTest.java [hunk 12]
```diff
@@ -0,0 +1,178 @@
...
+        assertSimilar(expected, result);
+    }
+
+    /**
+     * Should use date format from property
+     */
+    @Test
+    public void testCorrectCustomFormatProperty() {
+        // given
+        System.setProperty("IGNITE_JAVA_LOGGER_DATE_FORMAT", "yyyy/MM/dd HH:mm:ss");
```
Comment (sashapolo) [inline modules/core/src/test/java/org/apache/ignite/logger/java/JavaLoggerFormatterTest.java:110]: There's a `WithSystemProperty` annotation, I think it should be used instead (but current approach is also fine)

File: modules/core/src/main/java/org/apache/ignite/logger/java/JavaLoggerFormatter.java [hunk 13]
```diff
@@ -19,47 +19,72 @@
...
 
-            ex = "\n" + stackTrace;
+    @NotNull
+    static DateTimeFormatter resolveDateFormat() {
+        String customDateFormat = IgniteSystemProperties.getString(IGNITE_JUL_DATE_FORMAT_PROPERTY);
+        if (customDateFormat != null) {
+            try {
+                return DateTimeFormatter.ofPattern(customDateFormat).withZone(ZoneId.systemDefault());
+            } catch (Exception e) {
+                System.err.println("Incorrect JavaLogger date-time format: " + customDateFormat);
```
[the remark this answers is not in the document]
Comment (solveme) [reply modules/core/src/main/java/org/apache/ignite/logger/java/JavaLoggerFormatter.java:83]: I guess that throwing an exception might fail node startup, we don't need fail fast here

File: modules/core/src/main/java/org/apache/ignite/logger/java/JavaLoggerFormatter.java [hunk 14]
```diff
@@ -19,47 +19,72 @@
...
+            t.printStackTrace(new PrintWriter(sw));
+        }
 
-            String stackTrace = sw.toString();
+        return sw.toString();
+    }
 
-            ex = "\n" + stackTrace;
+    @NotNull
+    static DateTimeFormatter resolveDateFormat() {
```
[the remark this answers is not in the document]
Comment (solveme) [reply modules/core/src/main/java/org/apache/ignite/logger/java/JavaLoggerFormatter.java:77]: made it package to allow access for testing purpose

File: modules/core/src/main/java/org/apache/ignite/logger/java/JavaLoggerFormatter.java (hunk 7 above)
Comment (solveme) [reply modules/core/src/main/java/org/apache/ignite/logger/java/JavaLoggerFormatter.java:53]: indeed, will change it

File: modules/core/src/main/java/org/apache/ignite/internal/util/IgniteUtils.java (hunk 1 above)
Comment (solveme) [reply modules/core/src/main/java/org/apache/ignite/internal/util/IgniteUtils.java:562]: I thought about it also, but ISO_LOCAL_DATE_TIME doesn't print millis

File: modules/core/src/test/java/org/apache/ignite/logger/java/JavaLoggerFormatterTest.java (hunk 4 above)
Comment (solveme) [reply modules/core/src/test/java/org/apache/ignite/logger/java/JavaLoggerFormatterTest.java:43]: Agree, made out of habit

File: modules/core/src/test/java/org/apache/ignite/logger/java/JavaLoggerFormatterTest.java (hunk 3 above)
Comment (solveme) [reply modules/core/src/test/java/org/apache/ignite/logger/java/JavaLoggerFormatterTest.java:149]: This formatter was created for composing expected date format. I defined it to guard us from occasional change of date format inside JavaLoggerFormatter (e.g. you can't use code under test for testing this code). But seems, indeed, usage of IgniteUtils.ISO_DATE_FMT should be safe enough

File: modules/core/src/main/java/org/apache/ignite/logger/java/JavaLoggerFormatter.java (hunk 6 above)
Comment (solveme) [reply modules/core/src/main/java/org/apache/ignite/logger/java/JavaLoggerFormatter.java:44]: Will do

File: modules/core/src/main/java/org/apache/ignite/internal/util/IgniteUtils.java (hunk 1 above)
Comment (sashapolo) [reply modules/core/src/main/java/org/apache/ignite/internal/util/IgniteUtils.java:562]: Why do we need millis, since ISO_LOCAL_DATE_TIME prints nanoseconds?

File: modules/core/src/main/java/org/apache/ignite/logger/java/JavaLoggerFormatter.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (sashapolo) [reply modules/core/src/main/java/org/apache/ignite/logger/java/JavaLoggerFormatter.java:77]: oh, it behaves differently depending on the system variable, missed that

File: modules/core/src/main/java/org/apache/ignite/logger/java/JavaLoggerFormatter.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (sashapolo) [reply modules/core/src/main/java/org/apache/ignite/logger/java/JavaLoggerFormatter.java:83]: How will this failure look like then? Why do we print something to stderr? Is this behavior consistent with other loggers?

File: modules/core/src/main/java/org/apache/ignite/internal/util/IgniteUtils.java (hunk 1 above)
Comment (solveme) [reply modules/core/src/main/java/org/apache/ignite/internal/util/IgniteUtils.java:562]: Are you sure about nanos? From ISO_LOCAL_DATE_TIME javadoc:
The ISO date-time formatter that formats or parses a date-time without an offset, such as '2011-12-03T10:15:30'.

Anyway, millis are useful when you need to understand what happened first (there can be too many events in during whole second), but nanos would be an overkill.

Moreover proposed format yyyy-MM-dd'T'HH:mm:ss,SSS matches our default format used by ignite-log4j2:
<PatternLayout pattern="[%d{ISO8601}][%-5p][%t][%c{1}]%notEmpty{[%markerSimpleName]} %m%n"/>

2023-05-08T09:04:06,626 INFO  [pgrunner-#node-1] [o.a.i.i.IgniteKernal%node-1] Logger: Log4J2Logger

Comment (sashapolo) [reply modules/core/src/main/java/org/apache/ignite/internal/util/IgniteUtils.java:562]: > Are you sure about nanos?

yes, because it uses `ISO_LOCAL_TIME` internally, which states `One to nine digits for the nano-of-second. As many digits will be output as required`

Anyway, I won't insist, current approach is also fine, so it's up to you

**Review by sashapolo — APPROVED** [review]
